### ISSUE TITLE
Cycle in the constraint graph, involving Boost.Regex and Zlib

### ISSUE THREAD

**Issue by man-in-black382** [microsoft/vcpkg#33866]
Getting a Cmake warning: 

```
VCPKG/scripts/buildsystems/vcpkg.cmake:639 (_add_library):
  Cannot generate a safe runtime search path for target TargetName
  because there is a cycle in the constraint graph:

dir 0 is [/Users/Development/Build/vcpkg_installed/arm64-osx-dynamic/debug/lib]
      dir 1 must precede it due to runtime library [libz.dylib]
    dir 1 is [/Users/Development/Build/vcpkg_installed/arm64-osx-dynamic/lib]
      dir 0 must precede it due to runtime library [libboost_regex.dylib]
```

**Environment**
- OS: [macOS]
- Compiler: Apple Clang 
- arm64-osx-dynamic

1. Install OpenImageIO package, that in turn pulls in Zlib.
2. Install Boost package.
3. call `find_package(OpenImageIO REQUIRED)` and `find_package(Boost REQUIRED COMPONENTS regex)`
4. Link to `OpenImageIO::OpenImageIO` and `Boost::regex` targets.

**Expected behavior**
No warning should be produced

**Comment by dg0yt** [comment]
> dir 0 is [/Users/Development/Build/vcpkg_installed/arm64-osx-dynamic/debug/lib]
>       dir 1 must precede it due to runtime library [libz.dylib]
>     dir 1 is [/Users/Development/Build/vcpkg_installed/arm64-osx-dynamic/lib]
>       dir 0 must precede it due to runtime library [libboost_regex.dylib]

There seems to be an unexpected cross-linking between dylibs from debug into release or from release into debug.
Maybe with additional context the offending lib/port would be obvious.

**Comment by man-in-black382** [comment]
Where should O point my attention? To Zlib/Regex portfile.cmake? Somewhere else? Not sure where to start here

**Comment by dg0yt** [comment]
On linux, I would look at the relevant dylibs with `ldd`. Somewhere in this chain is the bad one, and the providing port must be fixed.

**Comment by man-in-black382** [comment]
Sorry for slow responses. Lots have piled up. Here is the analysis of the linked libs.

The two libraries in question: Zlib and Boost.Regex:

Regex DEBUG:

`Library: /Users/Development/Project/Build/macos-openfx-debug/vcpkg_installed/arm64-osx-dynamic/debug/lib/libboost_regex.dylib
Linked Libraries:
  @rpath/libboost_regex.dylib
  /usr/lib/libc++.1.dylib
  /usr/lib/libSystem.B.dylib
@rpath Values:`

Regex RELEASE:

`Library: /Users/Development/Project/Build/macos-openfx-debug/vcpkg_installed/arm64-osx-dynamic/lib/libboost_regex.dylib
Linked Libraries:
  @rpath/libboost_regex.dylib
  /usr/lib/libc++.1.dylib
  /usr/lib/libSystem.B.dylib
@rpath Values:`

Zlib DEBUG:

`Library: /Users/Development/Project/Build/macos-openfx-debug/vcpkg_installed/arm64-osx-dynamic/debug/lib/libz.dylib
Linked Libraries:
  @rpath/libz.1.dylib
  /usr/lib/libSystem.B.dylib
@rpath Values:`

Zlib RELEASE:

`Library: /Users/Development/Project/Build/macos-openfx-debug/vcpkg_installed/arm64-osx-dynamic/lib/libz.dylib
Linked Libraries:
  @rpath/libz.1.dylib
  /usr/lib/libSystem.B.dylib
@rpath Values:`

These even do not have rpaths set in them, so they cannot "go out" of their respective directories to cross-point to a dylib in another directory (debug->release, release->debug).

There were no dylibs found that would contain rpaths with ../.., that would allow it to go outside the debug/lib directory into lib and do ../debug that would allow release libs to point to debug libs. 

TL;DR I do not see cross links
